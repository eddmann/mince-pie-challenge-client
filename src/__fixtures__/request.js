export default function createRequestStack(responses = []) {
  const push = ({ path, method, body, response }) => responses.push({ path, method, body, response });

  const api = (path, method, body = undefined) => {
    const actual = { path, method, body };

    if (responses.length === 0) {
      return fail(`Expected: Nil, Actual: ${JSON.stringify(actual)}`);
    }

    const { response, ...expected } = responses.shift();

    expect(expected).toEqual(actual);

    return Promise.resolve(response);
  };

  const upload = (url, file) => api(url, 'PUT', file);

  return { push, api, upload };
};
