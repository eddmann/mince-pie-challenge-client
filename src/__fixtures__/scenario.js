export default function createScenario(fixtures = () => ({})) {
  return (label, run) => {
    const steps = [];

    run({
      given: (label, step) => steps.push(step),
      when: (label, step) => steps.push(step),
      then: (label, step) => steps.push(step),
      fixtures: fixtures(),
    });

    it(label, () =>
      steps.reduce((state, step) => state
        .then(wrapper => step(wrapper) || wrapper)
        .then(wrapper => new Promise(res => process.nextTick(() => { wrapper && wrapper.update(); res(wrapper); }))),
      Promise.resolve(undefined)));
  };
};
