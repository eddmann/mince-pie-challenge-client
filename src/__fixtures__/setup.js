import React from 'react';
import Enzyme, { mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import createMemoryHistory from 'history/createBrowserHistory';
import createScenario from './scenario';
import createRequestStack from './request';
import app from '../app';

Enzyme.configure({ adapter: new Adapter() });

export const scenario = createScenario(() => {
  const history = createMemoryHistory();
  const request = createRequestStack();
  const auth = { getToken: () => Promise.resolve() };

  return {
    request,
    auth,
    renderApp: (path = '/') => {
      history.replace(path);
      return app({
        history,
        services: { request, auth },
        links: { add: 'add.url', list: 'list.url', view: 'view.url/{id}' },
      }).then(mount);
    },
  };
});
