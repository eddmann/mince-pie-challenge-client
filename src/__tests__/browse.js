import { scenario } from '../__fixtures__/setup';

scenario('lists no mince pies', ({ given, when, then, fixtures: { request, renderApp } }) => {
  given('empty mince pie list resource', () => {
    request.push({
      path: 'list.url',
      method: 'GET',
      response: {
        _links: { self: { href: '/pies' } },
        _embedded: { pies: [] },
        total: 0,
      },
    });
  });

  when('visit the browse view', wrapper => renderApp('/'));

  then('no pies are present in listing', wrapper => {
    expect(wrapper.find('CardGroup')).toMatchSnapshot();
    expect(wrapper.find('Card').length).toEqual(0);
  });
});

scenario('lists several mince pies', ({ given, when, then, fixtures: { request, renderApp } }) => {
  given('list resource with several mince pies', () => {
    request.push({
      path: 'list.url',
      method: 'GET',
      response: {
        _links: { self: { href: '/pies' } },
        _embedded: {
          pies: [
            {
              _links: { self: { href: '/pies/1' } },
              id: '1',
              name: 'Awesome Mince Pie',
              thumbnail: 'thumbnail-1.url',
              rating: { avg: 4, total: 2 },
            },
            {
              _links: { self: { href: '/pies/2' } },
              id: '2',
              name: 'Super Mince Pie',
              thumbnail: 'thumbnail-2.url',
              rating: { avg: 3, total: 4 },
            },
          ],
        },
        total: 2,
      },
    });
  });

  when('visit the browse view', wrapper => renderApp('/'));

  then('several mince pies are present in listing', wrapper => {
    expect(wrapper.find('CardGroup')).toMatchSnapshot();
    expect(wrapper.find('Card').length).toEqual(2);
  });
});
