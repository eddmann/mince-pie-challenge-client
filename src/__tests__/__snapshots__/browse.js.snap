// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lists no mince pies 1`] = `
<CardGroup
  doubling={true}
  itemsPerRow={4}
>
  <div
    className="ui doubling four cards"
  />
</CardGroup>
`;

exports[`lists several mince pies 1`] = `
<CardGroup
  doubling={true}
  itemsPerRow={4}
>
  <div
    className="ui doubling four cards"
  >
    <Card
      className="Pie"
      fluid={true}
      id="1"
      key="1"
      onClick={[Function]}
    >
      <a
        className="ui fluid card Pie"
        id="1"
        onClick={[Function]}
      >
        <Image
          as="img"
          src="thumbnail-1.url"
          ui={true}
        >
          <img
            className="ui image"
            src="thumbnail-1.url"
          />
        </Image>
        <CardContent>
          <div
            className="content"
          >
            <CardHeader
              content="Awesome Mince Pie"
            >
              <div
                className="header"
              >
                Awesome Mince Pie
              </div>
            </CardHeader>
            <CardDescription>
              <div
                className="description"
              >
                <Rating
                  className="Rating"
                  clearable="auto"
                  defaultRating={4}
                  disabled={true}
                  maxRating={5}
                >
                  <div
                    className="ui disabled rating Rating"
                    onMouseLeave={[Function]}
                    role="radiogroup"
                  >
                    <RatingIcon
                      active={true}
                      aria-checked={false}
                      aria-posinset={1}
                      aria-setsize={5}
                      as="i"
                      index={0}
                      key="0"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={1}
                        aria-setsize={5}
                        className="active icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={true}
                      aria-checked={false}
                      aria-posinset={2}
                      aria-setsize={5}
                      as="i"
                      index={1}
                      key="1"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={2}
                        aria-setsize={5}
                        className="active icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={true}
                      aria-checked={false}
                      aria-posinset={3}
                      aria-setsize={5}
                      as="i"
                      index={2}
                      key="2"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={3}
                        aria-setsize={5}
                        className="active icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={true}
                      aria-checked={true}
                      aria-posinset={4}
                      aria-setsize={5}
                      as="i"
                      index={3}
                      key="3"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={true}
                        aria-posinset={4}
                        aria-setsize={5}
                        className="active icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={false}
                      aria-checked={false}
                      aria-posinset={5}
                      aria-setsize={5}
                      as="i"
                      index={4}
                      key="4"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={5}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                  </div>
                </Rating>
              </div>
            </CardDescription>
          </div>
        </CardContent>
      </a>
    </Card>
    <Card
      className="Pie"
      fluid={true}
      id="2"
      key="2"
      onClick={[Function]}
    >
      <a
        className="ui fluid card Pie"
        id="2"
        onClick={[Function]}
      >
        <Image
          as="img"
          src="thumbnail-2.url"
          ui={true}
        >
          <img
            className="ui image"
            src="thumbnail-2.url"
          />
        </Image>
        <CardContent>
          <div
            className="content"
          >
            <CardHeader
              content="Super Mince Pie"
            >
              <div
                className="header"
              >
                Super Mince Pie
              </div>
            </CardHeader>
            <CardDescription>
              <div
                className="description"
              >
                <Rating
                  className="Rating"
                  clearable="auto"
                  defaultRating={3}
                  disabled={true}
                  maxRating={5}
                >
                  <div
                    className="ui disabled rating Rating"
                    onMouseLeave={[Function]}
                    role="radiogroup"
                  >
                    <RatingIcon
                      active={true}
                      aria-checked={false}
                      aria-posinset={1}
                      aria-setsize={5}
                      as="i"
                      index={0}
                      key="0"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={1}
                        aria-setsize={5}
                        className="active icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={true}
                      aria-checked={false}
                      aria-posinset={2}
                      aria-setsize={5}
                      as="i"
                      index={1}
                      key="1"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={2}
                        aria-setsize={5}
                        className="active icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={true}
                      aria-checked={true}
                      aria-posinset={3}
                      aria-setsize={5}
                      as="i"
                      index={2}
                      key="2"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={true}
                        aria-posinset={3}
                        aria-setsize={5}
                        className="active icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={false}
                      aria-checked={false}
                      aria-posinset={4}
                      aria-setsize={5}
                      as="i"
                      index={3}
                      key="3"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={4}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                    <RatingIcon
                      active={false}
                      aria-checked={false}
                      aria-posinset={5}
                      aria-setsize={5}
                      as="i"
                      index={4}
                      key="4"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                      selected={false}
                    >
                      <i
                        aria-checked={false}
                        aria-posinset={5}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                      />
                    </RatingIcon>
                  </div>
                </Rating>
              </div>
            </CardDescription>
          </div>
        </CardContent>
      </a>
    </Card>
  </div>
</CardGroup>
`;
